From 9b195569201a77a23d7a80491d672ab64e08bf68 Mon Sep 17 00:00:00 2001
From: zhangzujian <zhangzujian.7@gmail.com>
Date: Thu, 6 Aug 2026 11:00:00 +0800
Subject: [PATCH] Avoid SIGPIPE when a control client disconnects

Control clients may disconnect while the beacon is preparing or sending a
multi-line reply. Prevent that connection-local failure from terminating the
daemon, and stop sending the remainder of the reply after the first failure.

Signed-off-by: zhangzujian <zhangzujian.7@gmail.com>
---
 CommandProcessor.cpp | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/CommandProcessor.cpp b/CommandProcessor.cpp
index f6f9191..2d61561 100755
--- a/CommandProcessor.cpp
+++ b/CommandProcessor.cpp
@@ -27,3 +27,4 @@ protected:
   Socket m_listenSocket;
   Socket m_replySocket;
+  bool m_replyFailed;
   RecvMsg m_inCommand;
@@ -50,3 +51,4 @@ public:
      m_listenSocket(),
      m_replySocket(),
+     m_replyFailed(false),
      m_inCommand(MaxCommandSize, 0),
@@ -327,2 +329,3 @@ protected:
     m_replySocket = connectedSocket; // note connectedSocket still 'owns' the socket.
+    m_replyFailed = false;
     m_replySocket.SetLogName(connectedSocket.LogName());
@@ -2053,5 +2056,8 @@ protected:
   void doMessageReply(const char *reply, size_t length)
   {
-    // TODO timeout? Check for shutdown?
-    m_replySocket.Send(reply, length);
+    if (m_replyFailed)
+      return;
+
+    if (!m_replySocket.Send(reply, length, MSG_NOSIGNAL))
+      m_replyFailed = true;
   }
